int32 num
int32 center_x
int32 center_y
bool flag